Enable Declarative Validation for ReplicationController

jpbetz opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR:

- **Enables validation-gen for core v1 group:** `+k8s:validation-gen` tags in `pkg/apis/core/v1/doc.go` enable generation for the core `v1` API group. When enabled `validation-gen` parses "+" tags i `staging/src/k8s.io/api/core/v1/types.go` and generates validation code into `pkg/apis/core/v1/zz_generated.validations.go`.

- **Updates `ReplicationController` strategy to run generated declarative validation:** When the `DeclarativeValidation` gate is enabled, the strategies' validate functions will execute both the existing hand-written validation logic and the newly generated declarative validation:
    - The validation errors returned by both methods are compared.
    - If the errors do not match for validations that are intended to be migrated to declarative, a metric named `declarative_validation_mismatch_total` is incremented.
    - If the declarative validation code encounters a panic, a metric named `declarative_validation_panic_total` is incremented.
   - Validation errors reported back to the caller respect the `DeclarativeValidationTakeover` feature gate:
     - For validation rules that have been migrated to declarative: When `DeclarativeValidationTakeover` is enabled, validation errors produced by declarative validation are returned. Otherwise, hand-written validation errors are returned. If declarative validation panics, an additional internal error is returned to report the problem.
     - For validation rules that have not yet been migrated, the existing hand-written validation errors will continue to be returned.

- **Enables declarative validation testing:** Declarative validation is performed on API versions (not on the internal version like hand written validation). Fuzz testing is enabled for the core `v1` group's validation.  Validation tests in `declarative_validation_test.go` are updated to use `apitesting.VerifyVersionedValidationEquivalence` to validate the (declarative) versioned validation.

- **Does NOT migrate any validations:** Validation migrations are introduced in a subsequent PR:
  - https://github.com/kubernetes/kubernetes/pull/130725





#### Special notes for your reviewer:

This PR is currently stacked on https://github.com/kubernetes/kubernetes/pull/130705.  Please ignore the commits from that PR when reviewing this PR.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

thockin committed:
Enable validation-gen on core/v1

This enables generation but does not activate the strategy calling of
declarative.

jpbetz commented:
/assign @deads2k 
This is a small PR that plays out the decisions made by https://github.com/kubernetes/kubernetes/pull/130705 and prepares the codebase for migrations of actual validations, which happen in https://github.com/kubernetes/kubernetes/pull/130725

thockin commented:
FWIW this is what I expected to see

aaron-prindle commented:
/triage accepted

thockin commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -16659,9 +16671,11 @@ func TestValidateReplicationControllerUpdate(t *testing.T) {
 		tc.old.ObjectMeta.ResourceVersion = "1"
 		tc.update.ObjectMeta.ResourceVersion = "1"
-		if errs := ValidateReplicationControllerUpdate(&tc.update, &tc.old, PodValidationOptions{}); len(errs) != 0 {
+		errs := ValidateReplicationControllerUpdate(&tc.update, &tc.old, PodValidationOptions{})
```
curious why this change?

jpbetz replied:
When I picked up commits, this is what we I got from the dev branch: https://github.com/jpbetz/kubernetes/blob/21600930a778dd13decc388f5e4ed956e8c31439/pkg/apis/core/validation/validation_test.go#L16640

I personally prefer the `if` clause, but I biased toward consistency with the dev branch.

thockin replied:
I t looks like it came from a series of changes which were made and removed (the gate checking that used to be here).  I will send a fix to dev, but we can pick it up later when we do stragglers.

jpbetz replied:
OK, After consolidating the group specific testing to strategy and the generic testing to pkg/api/testing I don't need any changes to validation_test.go so I'm dropping it from this PR.

thockin commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -16866,6 +16885,107 @@ func TestValidateReplicationController(t *testing.T) {
+}
+
+// FIXME: move somewhere generic - pkg/api/testing?
+func TestVersionedValidationByFuzzing(t *testing.T) {
```
This is OK in this pkg, but it doesn't feel ideal.  Fast-forward to the day when there's no manual validation, all the testing is through strategy, where does this belong?  pkg/registry/core?  pkg/registry?

jpbetz replied:
I've moved it into pkg/api/testing adjacent to generic testing for defaulting.

thockin commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -16866,6 +16885,107 @@ func TestValidateReplicationController(t *testing.T) {
+	}
+}
+
+func verifyVersionedValidationEquivalence(t *testing.T, obj, old k8sruntime.Object) {
```
I mentioned in the other PR, I think, but this feels like it should run against strategy (where we are aware of versioned validation) rather than this pkg?

jpbetz replied:
Sorry, yes.  I'll incorporate.

jpbetz replied:
OK, this is now resolved. The testing is now consolidated to strategy_test (for group specific testing), and pkg/api/testing (for generic testing like fuzzing).

jpbetz commented:
/retest

jpbetz commented on pkg/registry/core/replicationcontroller/strategy.go:
```diff
@@ -174,6 +196,26 @@ func (rcStrategy) ValidateUpdate(ctx context.Context, obj, old runtime.Object) f
+		takeover := utilfeature.DefaultFeatureGate.Enabled(features.DeclarativeValidationTakeover)
+
+		// Run both validations with panic recovery
+		declarativeErrs := rest.ValidateDeclarativelyWithRecovery(ctx, nil, legacyscheme.Scheme, obj, takeover)
```
Apologies. Wires got cross on our side.

Only the validateUpdate call is needed and it [works with ratcheting](https://github.com/kubernetes/kubernetes/blob/1c756849d627dfb352eda2f3a0664eadbd488a0a/staging/src/k8s.io/apimachinery/pkg/api/validate/doc.go#L39-L45) (we may add utility tags to make ratcheting easier too).

The way declarative validation works is we funnel both create and update requests into the same validation code.  So by default, a call to "validate update" calls all validations.  It would only be if a validation did special a special check like `oldObj != nil` or `operation.Type == Create` that a validation would be skipped.  We don't have any validations that do that.

I've simplified this to the way it should be.

Thanks for catching.

jpbetz replied:
I can offer a [demo PR](github.com/kubernetes/kubernetes/pull/130753) that shows ratcheting working as intended. [It includes tests showing the ratcheting working as expected](https://github.com/kubernetes/kubernetes/pull/130753/commits/7fd1e3b84b4af71236b9679fd31fccbd0751e21e#diff-883b42cb2642e4c4b94cdb94f18c1c62c3992d62b3dfa5309f9ec569e059dfd3R340-R397).

Can I offer an IOU for in-tree testing?  We can have that in 1.34 I think.

It's a bit tricky to set up because, during the migration, we require that we have handwritten validation corresponding to all declarative validation, so I had to add a hand written validation example.

We also don't have any existing ratcheting validation tags, and there are a lot of design questions there that I'd like to make sure we resolve fully.  Do we do:

```
+k8s:maxLength=64
+k8s:tighten=+k8s:maxLength=32
```

Or do we do something like:

```
+k8s:ratchet=+k8s:maxLength=64,+k8s:maxLength=32
```

Or do we have specialized tags (my demo just uses a `k8s:tightenedMinimum` tag to show the idea).

We do have [testing for `+k8s:immutable`](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/tags/immutable) which shows how the how update validation handles the parallel traverse of "object" and "oldObject"

thockin replied:
I mostly want to be careful that we don't over-drive the headlights.  I am confident we can do ratcheting, but I want to have a couple of concrete cases in mind when we do.  This is something we can seek out in the 1.34 window, for sure.

jpbetz commented:
Rebased onto master.

jpbetz committed:
ReplicationController: Enable declarative validation

After declarative validation is enabled in the ReplicationController
strategy in this way, the generated declarative validation code
in pkg/apis/core/v1/zz.generated.validations.go will be run
when the strategy validates ReplicationController.

Co-authored-by: Tim Hockin <thockin@google.com>
Co-authored-by: Aaron Prindle <aprindle@google.com>
Co-authored-by: Yongrui Lin <yongrlin@google.com>
Co-authored-by: David Eads <deads@redhat.com>

thockin commented on pkg/registry/core/replicationcontroller/strategy_test.go:
```diff
@@ -205,3 +209,216 @@ func TestValidateUpdate(t *testing.T) {
+		APIGroup:   "",
+		APIVersion: "v1",
+	})
+	successCases := []api.ReplicationController{
```
This is kind of a nit, but also maybe important since this is the exemplar:

Do we want every such test to keep success and failure cases in two different loops?  

What I am suggesting, concretely is:

```diff
diff --git a/pkg/registry/core/replicationcontroller/strategy_test.go b/pkg/registry/core/replicationcontroller/strategy_test.go
index 9779211d5f1..34cc284ae05 100644
--- a/pkg/registry/core/replicationcontroller/strategy_test.go
+++ b/pkg/registry/core/replicationcontroller/strategy_test.go
@@ -236,39 +236,29 @@ func TestValidateForDeclarative(t *testing.T) {
 		APIGroup:   "",
 		APIVersion: "v1",
 	})
-	successCases := []api.ReplicationController{
-		mkValidReplicationController(func(rc *api.ReplicationController) {}),
-		mkValidReplicationController(func(rc *api.ReplicationController) { rc.Spec.Replicas = 0 }),
-		mkValidReplicationController(func(rc *api.ReplicationController) { rc.Spec.Replicas = 1 }),
-		mkValidReplicationController(func(rc *api.ReplicationController) { rc.Spec.Replicas = 100 }),
-		mkValidReplicationController(func(rc *api.ReplicationController) { rc.Spec.MinReadySeconds = 0 }),
-		mkValidReplicationController(func(rc *api.ReplicationController) { rc.Spec.MinReadySeconds = 1 }),
-		mkValidReplicationController(func(rc *api.ReplicationController) { rc.Spec.MinReadySeconds = 100 }),
-	}
-	for _, tc := range successCases {
-		for _, gateVal := range []bool{true, false} {
-			// We only need to test both gate enabled and disabled together, because
-			// 1) the DeclarativeValidationTakeover won't take effect if DeclarativeValidation is disabled.
-			// 2) the validation output, when only DeclarativeValidation is enabled, is the same as when both gates are disabled.
-			featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.DeclarativeValidation, gateVal)
-			featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.DeclarativeValidationTakeover, gateVal)
-			errs := Strategy.Validate(ctx, &tc)
-			if len(errs) != 0 {
-				t.Errorf("expected success: %v", errs)
-			}
-		}
-	}
 
-	errorCases := map[string]struct {
+	testCases := map[string]struct {
 		input        api.ReplicationController
 		expectedErrs field.ErrorList
 	}{
+		"0 replicas": {
+			input: mkValidReplicationController(func(rc *api.ReplicationController) { rc.Spec.Replicas = 0 }),
+		},
+		"positive replicas": {
+			input: mkValidReplicationController(func(rc *api.ReplicationController) { rc.Spec.Replicas = 3 }),
+		},
 		"negative replicas": {
 			input: mkValidReplicationController(func(rc *api.ReplicationController) { rc.Spec.Replicas = -1 }),
 			expectedErrs: field.ErrorList{
 				field.Invalid(field.NewPath("spec.replicas"), nil, "").WithOrigin("minimum"),
 			},
 		},
+		"0 minReadySeconds": {
+			input: mkValidReplicationController(func(rc *api.ReplicationController) { rc.Spec.MinReadySeconds = 0 }),
+		},
+		"positive minReadySeconds": {
+			input: mkValidReplicationController(func(rc *api.ReplicationController) { rc.Spec.MinReadySeconds = 3 }),
+		},
 		"negative minReadySeconds": {
 			input: mkValidReplicationController(func(rc *api.ReplicationController) { rc.Spec.MinReadySeconds = -1 }),
 			expectedErrs: field.ErrorList{
@@ -276,7 +266,7 @@ func TestValidateForDeclarative(t *testing.T) {
 			},
 		},
 	}
-	for k, tc := range errorCases {
+	for k, tc := range testCases {
 		t.Run(k, func(t *testing.T) {
 			var declarativeTakeoverErrs field.ErrorList
 			var imperativeErrs field.ErrorList
@@ -313,51 +303,23 @@ func TestValidateUpdateForDeclarative(t *testing.T) {
 		APIVersion: "v1",
 	})
 
-	successCases := []struct {
-		old    api.ReplicationController
-		update api.ReplicationController
-	}{{
-		old: mkValidReplicationController(func(rc *api.ReplicationController) {}),
-		update: mkValidReplicationController(func(rc *api.ReplicationController) {
-			rc.Spec.Replicas = 0
-		}),
-	}, {
-		old: mkValidReplicationController(func(rc *api.ReplicationController) {}),
-		update: mkValidReplicationController(func(rc *api.ReplicationController) {
-			rc.Spec.Replicas = 3
-		}),
-	}, {
-		old: mkValidReplicationController(func(rc *api.ReplicationController) {}),
-		update: mkValidReplicationController(func(rc *api.ReplicationController) {
-			rc.Spec.MinReadySeconds = 0
-		}),
-	}, {
-		old: mkValidReplicationController(func(rc *api.ReplicationController) {}),
-		update: mkValidReplicationController(func(rc *api.ReplicationController) {
-			rc.Spec.MinReadySeconds = 3
-		}),
-	}}
-	for _, tc := range successCases {
-		tc.old.ObjectMeta.ResourceVersion = "1"
-		tc.update.ObjectMeta.ResourceVersion = "1"
-		for _, gateVal := range []bool{true, false} {
-			// We only need to test both gate enabled and disabled together, because
-			// 1) the DeclarativeValidationTakeover won't take effect if DeclarativeValidation is disabled.
-			// 2) the validation output, when only DeclarativeValidation is enabled, is the same as when both gates are disabled.
-			featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.DeclarativeValidation, gateVal)
-			featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.DeclarativeValidationTakeover, gateVal)
-			errs := Strategy.ValidateUpdate(ctx, &tc.update, &tc.old)
-			if len(errs) != 0 {
-				t.Errorf("expected success: %v", errs)
-			}
-		}
-	}
-
 	errorCases := map[string]struct {
 		old          api.ReplicationController
 		update       api.ReplicationController
 		expectedErrs field.ErrorList
 	}{
+		"0 replicas": {
+			old: mkValidReplicationController(func(rc *api.ReplicationController) {}),
+			update: mkValidReplicationController(func(rc *api.ReplicationController) {
+				rc.Spec.Replicas = 0
+			}),
+		},
+		"positive replicas": {
+			old: mkValidReplicationController(func(rc *api.ReplicationController) {}),
+			update: mkValidReplicationController(func(rc *api.ReplicationController) {
+				rc.Spec.Replicas = 3
+			}),
+		},
 		"negative replicas": {
 			old: mkValidReplicationController(func(rc *api.ReplicationController) {}),
 			update: mkValidReplicationController(func(rc *api.ReplicationController) {
@@ -367,6 +329,18 @@ func TestValidateUpdateForDeclarative(t *testing.T) {
 				field.Invalid(field.NewPath("spec.replicas"), nil, "").WithOrigin("minimum"),
 			},
 		},
+		"0 minReadySeconds": {
+			old: mkValidReplicationController(func(rc *api.ReplicationController) {}),
+			update: mkValidReplicationController(func(rc *api.ReplicationController) {
+				rc.Spec.MinReadySeconds = 0
+			}),
+		},
+		"positive minReadySeconds": {
+			old: mkValidReplicationController(func(rc *api.ReplicationController) {}),
+			update: mkValidReplicationController(func(rc *api.ReplicationController) {
+				rc.Spec.MinReadySeconds = 3
+			}),
+		},
 		"negative minReadySeconds": {
 			old: mkValidReplicationController(func(rc *api.ReplicationController) {}),
 			update: mkValidReplicationController(func(rc *api.ReplicationController) {
```

jpbetz commented on pkg/registry/core/replicationcontroller/strategy.go:
```diff
@@ -24,6 +24,8 @@ import (
 	"strconv"
 	"strings"
 
+	"sigs.k8s.io/structured-merge-diff/v4/fieldpath"
```
Heh, Long story.. my computer broke and I had to get a new one and I had to reconfigure my IDEs import grouping rules but I got it wrong and didn't realize it untl just now :neutral_face: 

Fixed now (both the import and my IDE settings).

jpbetz commented on pkg/registry/core/replicationcontroller/strategy.go:
```diff
@@ -202,10 +203,7 @@ func (rcStrategy) ValidateUpdate(ctx context.Context, obj, old runtime.Object) f
-		declarativeErrs = append(
-			declarativeErrs,
-			rest.ValidateUpdateDeclarativelyWithRecovery(ctx, nil, legacyscheme.Scheme, obj, old, takeover)...)
+		declarativeErrs := rest.ValidateUpdateDeclarativelyWithRecovery(ctx, nil, legacyscheme.Scheme, newRc, oldRc, takeover)
```
SG, Added a comment.

jpbetz commented:
@thockin Feedback applied in separate commit.

jpbetz commented on pkg/registry/core/replicationcontroller/strategy_test.go:
```diff
@@ -205,3 +209,206 @@ func TestValidateUpdate(t *testing.T) {
+	return rc
+}
+
+func TestValidateForDeclarative(t *testing.T) {
```
@thockin We could take the organization one step farther and put these tests into a file `validation_test.go` file in the this registry directory.  The way these test are now we're appending validation tests to other non-related strategy tests...

Worth doing?

thockin replied:
That sounds like a reasonable thing to do before they start getting large

jpbetz replied:
Pushed in a separate commit.  Tell me if you hate it.

thockin reviewed (commented):
Looks clean, but I do think you should fixup the last commits into the smallest possible example PR

jpbetz committed:
Enable versioned validation fuzz testing for core v1 group

jpbetz commented on pkg/registry/core/replicationcontroller/declarative_validation_test.go:
```diff
@@ -0,0 +1,234 @@
+		update       api.ReplicationController
+		expectedErrs field.ErrorList
+	}{
+		"0 replicas": {
```
Well, I guess we could argue that organizing the tests made this more obvious.  It's it's a least a small win.

jpbetz commented:
> Looks clean, but I do think you should fixup the last commits into the smallest possible example PR

Reorganized into a clean set of minimal commits.

jpbetz committed:
ReplicationController: Add declarative validation test suite

Introduce a test suite that ensures declarative test cases
are fully tested and that validation errors are compared
with handwritten validation to ensure consistency.

Co-authored-by: Tim Hockin <thockin@google.com>
Co-authored-by: Aaron Prindle <aprindle@google.com>
Co-authored-by: Yongrui Lin <yongrlin@google.com>

jpbetz commented:
Declarative validations tests of individual fields have been relocated to https://github.com/kubernetes/kubernetes/pull/130725 to reside alongside commits that enable declarative validation of those fields.

jpbetz commented:
/retest

thockin commented:
Thanks!

/lgtm
/approve
